source/logrepl/internal/subscription.go

```diff
@@ -214,9 +218,31 @@ func (s *Subscription) handleXLogData(ctx context.Context, copyDataMsg *pgproto3
 		return fmt.Errorf("handler error: %w", err)
 	}
 
-	if xld.WALStart > 0 {
-		s.walWritten = xld.WALStart
+	// Track `BEGIN` and `COMMIT` messages. Track the expected end LSN.
+	// Only update the last written LSN to data changing messages (update, insert, delete)
+	switch m := logicalMsg.(type) {
+	case *pglogrepl.BeginMessage:
+		s.lastTXLSN = m.FinalLSN
+
+		sdk.Logger(ctx).Debug().
+			Stringer("final_lsn", m.FinalLSN).
+			Msg("received begin msg")
+	case *pglogrepl.CommitMessage:
+		sdk.Logger(ctx).Debug().
+			Stringer("commit_lsn", m.CommitLSN).
+			Stringer("tx_lsn", m.TransactionEndLSN).
+			Stringer("expected_end_lsn", s.lastTXLSN).
+			Msg("received commit")
+
+		if s.lastTXLSN != 0 && s.lastTXLSN != m.CommitLSN {
+			return fmt.Errorf("out of order commit %s, expected %s", m.CommitLSN, s.lastTXLSN)
+		}
+	default:
+		if xld.WALStart > 0 {
+			s.walWritten = xld.WALStart
```

Comment:
Doesn't this part potentially mess up the check on line 352? In case when `s.Handler` in line 217 doesn't create a record, we will have a LSN in `s.walWritten` that will never match `s.walFlushed`.

If `s.Handler` in line 217 would return a boolean, telling us if it created a record or not, we could decide if this LSN should be written into `s.walWritten` or into `s.lastTXLSN`. In other words - we would be tracking the LSN of any message that requires an ack in `s.walWritten`, while we would be tracking the largest LSN we have received of any other message in `s.lastTXLSN`. If we do that, the naming doesn't fit that well anymore though.

Reply:
> Doesn't this part potentially mess up the check on line 352? In case when s.Handler in line 217 doesn't create a record, we will have a LSN in s.walWritten that will never match s.walFlushed.

Only if we are not covering all the messages, but for proto v1 we (mostly ) are. The handler will deal with the changes and relation messages and here we deal with the begin/commit to track the tx end LSN to ensure we receive it. 

> If s.Handler in line 217 would return a boolean, telling us if it created a record or not, we could decide if this LSN should be written into s.walWritten or into s.lastTXLSN. In other words - we would be tracking the LSN of any message that requires an ack in s.walWritten, while we would be tracking the largest LSN we have received of any other message in s.lastTXLSN. If we do that, the naming doesn't fit that well anymore though.

Handler returning a signal when the message has been written makes sense, but let's not confuse the meaning of what `walWritten` is compared to begin message end of tx lsn. 

This can change such that the handler can return an error when it hasn't encountered the commit of the stream its processing. Essentially pushing the lastTXLSN in the handler code.

Will takes makes sense:
1. Change Handler to return bool or LSN which described what was written
2. Track begin/commit inside the handler struct

For context

https://www.postgresql.org/docs/current/protocol-replication.html#PROTOCOL-REPLICATION-START-REPLICATION

https://www.postgresql.org/docs/current/protocol-logicalrep-message-formats.html

Reply:
@lovromazgon addressed here 55f4d7b7eebb98d9143f62c2fbb4767f39983f5b